### ISSUE TITLE
custom-op, shift and the inference accuracy drop

### ISSUE THREAD

**Issue by Eunhui-Kim** [tensorpack/tensorpack#1462]
My tensorflow v 1.15.3
 I built custom-op active-shift-ops as this [url ](https://github.com/tensorflow/custom-op ) guides. 
 OS ubuntu 16.04
 gcc 7.5.0
 python 3.7 

 Thank you for useful tensorpack.
 It's available to me. 
 The problem is that the resnet training accuracy with my new custom-op, shift works fine, 
 however the accuracy of test is poor unexpected unser 30%. 

 I've make custom-op based on active-shift tensorflow github code. 
 And I can check that the custom operation works fine. 
 I check the basic test operation of shift.

My System Environment is as follows.
Python 3.7.8 (default, Jun 29 2020, 05:46:05)
[GCC 5.4.0 20160609] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import tensorflow as tf
>>> print(tf.__version__)
1.15.3
>>> import tensorpack as tp
2020-07-13 17:29:05.581202: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcudart.so.10.0
>>> print(tp.__version__)
0.10.1
>>> from tensorflow_active_shift.python.ops import active_shift2d_ops
>>>

 By using the tensorpack's example code, I can evaluate the cifar10-preact18.py.
 And I substitute the Spatial Conv2D with Shift Operation as follows.
-residual = Conv2D('conv2', residual, num_filters, kernel_size=3, strides=1, use_bias=False)
+residual = Shift2D('shift1', net, n_shifts, num_filters_in, stride=stride, padding='SAME')

and I made shift2d.py similar to conv2d.py as follows.

`
from tensorflow_active_shift.python.ops import active_shift2d_ops
@layer_register(log_shape=True)
@convert_to_tflayer_args(
    args_names=['filters','kernel_size'],
    name_mapping={
        'out_channel': 'filters',
        'kernel_shape': 'kernel_size',
        'strides': 'stride',
        'paddings' : 'padding'
    })

def Shift2D(
        inputs,
        filters, #n_shifts=2,     #number of shifts position to compute
        kernel_size, #shifts_size,  #number of channels
        stride=1,
        padding='SAME', #'same',
        data_format='NCHW',
	normalize=True):
....

filter_shape = (filters, kernel_size)
    input_dtype = inputs.dtype
W = tf.get_variable('W', filter_shape, dtype=input_dtype, initializer = tf.variance_scaling_initializer(scale=1.0 / 3, mode='fan_in', distribution='uniform'))

    with tf.variable_scope("shift"):
            layer = active_shift2d_ops.active_shift2d_op(
                inputs,
                shift=shift_tensor,
                strides=strides,
                paddings=paddings,
                data_format=data_format) 
    ret = tf.identity(layer, name='output')
    ret.variables = VariableHolder(W=W)
    return ret
`

At the first time, since InferenceGraph doesn't include Shift custom-op. 
Thus after I add tf.disable_eager_execution() into inference_runner.py and saver.py 
I can check the InferenceGraph also include the shift operation. 

`
import tensorflow as tf
from tensorflow_active_shift.python.ops import active_shift2d_ops
from tensorflow.python.saved_model import builder as saved_model_builder
from tensorflow.python.saved_model import signature_constants, signature_def_utils, tag_constants, utils
from tensorflow.python.util import compat
import os

config = tf.ConfigProto(allow_soft_placement=True)
sess = tf.Session(config=config)

saver = tf.train.import_meta_graph('./train_log/cifar10-Shift20/graph-0710-183700.meta')
saver.restore(sess, tf.train.latest_checkpoint('./train_log/cifar10-Shift20/'))
graph = tf.get_default_graph()
model_signature = tf.saved_model.signature_def_utils.predict_signature_def(
    inputs={'image':graph.get_tensor_by_name('input:0')}, 
    outputs={'label':graph.get_tensor_by_name('label:0')})
for op in graph.get_operations():
    print(op.name)
....
InferenceTower/res2.2/conv1/output
InferenceTower/res2.2/shift1/shift/shift/shift
InferenceTower/res2.2/shift1/shift/shift
.....
`

Is there any idea to solve this problem?

Thank you in advance ~

**Comment by ppwwyyxx** [comment]
Please check the unmodified example to see if there are any problems with training or test output. If so, please provide details following the issue template (click "New Issue" -> "Unexpected Problems / Bugs", or visit [this link to post issues about unexpected problems](https://github.com/tensorpack/tensorpack/issues/new?template=unexpected-problems---bugs.md)).
If the error is low only when using your custom op, it's unlikely a tensorpack issue and not something we can help with.

**Comment by Eunhui-Kim** [comment]
OK, I'll seek another part. 
After I found solution, I'll close. 
Thank you for your clue~

**Comment by Eunhui-Kim** [comment]
In tensorpack,
   conv2d.py file,
   ```
   with rename_get_variable({'kernel': 'W', 'bias': 'b'}):
            layer = tf.layers.Conv2D(
                filters,
                kernel_size,
                strides=strides,
                padding=padding,
                data_format=data_format,
                dilation_rate=dilation_rate,
                activation=activation,
                use_bias=use_bias,
                kernel_initializer=kernel_initializer,
                bias_initializer=bias_initializer,
                kernel_regularizer=kernel_regularizer,
                bias_regularizer=bias_regularizer,
                activity_regularizer=activity_regularizer,
                _reuse=tf.get_variable_scope().reuse)
            ret = layer.apply(inputs, scope=tf.get_variable_scope())
            ret = tf.identity(ret, name='output')

        ret.variables = VariableHolder(W=layer.kernel)
   ```
  There is no `W` weight kernel_shape. but Variable Holder keep W.
 
  in another part when split, and dilation_rate is over 1
  ```
        kernel_shape = shape2d(kernel_size)
        filter_shape = kernel_shape + [in_channel / split, out_channel]
        stride = shape4d(strides, data_format=data_format)
        W = tf.get_variable('W', filter_shape, dtype=inputs_dtype, initializer=kernel_initializer)
  ``` 
 The filter shape might include [kernel_size and in_channel and out_channel]

 I made similarly shift2d.py
 
```
from tensorflow_active_shift.python.ops import active_shift2d_ops
@layer_register(log_shape=True)
@convert_to_tflayer_args(
args_names=['filters','kernel_size'],
name_mapping={
'out_channel': 'filters',
'kernel_shape': 'kernel_size',
'strides': 'stride',
'paddings' : 'padding'
})

def Shift2D(
inputs,
filters, #n_shifts=2, #number of shifts position to compute
kernel_size, #shifts_size, #number of channels
stride=1,
padding='SAME', #'same',
data_format='NCHW',
normalize=True):
....

filter_shape = (filters, kernel_size)
input_dtype = inputs.dtype
W = tf.get_variable('W', filter_shape, dtype=input_dtype, initializer = tf.variance_scaling_initializer(scale=1.0 / 3, mode='fan_in', distribution='uniform'))
with tf.variable_scope("shift"):
        layer = active_shift2d_ops.active_shift2d_op(
            inputs,
            shift=W,
            strides=strides,
            paddings=paddings,
            data_format=data_format) 
ret = tf.identity(layer, name='output')
ret.variables = VariableHolder(W=W)
return ret
```

I think it's the reason to show different accuracy...
How should I make the kernel shape?
Could you give me the guide to get correct accuracy?

Thank you in advance

**Comment by ppwwyyxx** [comment]
I'm not sure what the question is about. If you have questions how to implement a tensorflow layer, it is unrelated to tensorpack and we cannot help with that. The tensorpack-specific parts like `convert_to_tflayer_args`, `layer_register`, `VariableHolder` can all be removed without affecting how it works (https://tensorpack.readthedocs.io/tutorial/symbolic.html)

**Comment by Eunhui-Kim** [comment]
OK!

tensorpack is perfect!

I found the solution.

I had to understand get_variable, and variable_scope better.

The Tip is to understand the error of my code, I have to use jupyter notebook to analyze the py file.

Thank you~
